docs/snapshotting/snapshot-support.md

```diff
@@ -0,0 +1,277 @@
+- the guest memory,
+- the emulated HW state (both KVM and Firecracker emulated HW),
+- the guest's storage disks' contents.
+
+The state of the components listed above is generated independently, which brings
+flexibility to our snapshotting support. This means that taking a snapshot results
+in multiple files that are composing the full microVM snapshot:
+- the guest memory file,
+- the microVM state file,
+- zero or more disk files (depending on how many the guest had; these are
+  **managed by the users**, which means they need to externally back up their
+  block devices backing files).
+
+This design allows sharing any read-only resources (such as read-only disks or
+microVM state) between multiple microVMs in order to improve density.
+When loading a snapshot, instead of loading at resume time the full contents from
+file to memory, Firecracker creates a `MAP_PRIVATE` mapping of the memory file,
+resulting in runtime on-demand loading of memory pages. Any subsequent memory writes
+go to a copy-on-write anonymous memory mapping.
+This has the advantage of very fast snapshot loading times, but comes with the cost
+of having to keep the guest memory file around for the entire lifetime of the
+resumed microVM.
+
+## Firecracker Snapshotting characteristics
+
+- Fresh Firecracker microVMs are booted using `anonymous` memory, while microVMs
+  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
+  to anonymous memory.
+- Resuming from a snapshot is optimized for speed. Taking a snapshot always involves
+  synchronously writing dirty memory pages to the memory snapshot file. Taking a
+  snapshot of a fresh microVM running off non-file-backed anonymous memory, results in
+  full contents of guest memory being written to the snapshot.
+- The snapshot operations have minimal requirements/dependencies on when/how they
+  should be used.
+- The API calls exposing the snapshotting functionality have clear **Prerequisites** that
+  describe the previously mentioned minimal requirements.
+- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
+  creation and the disk contents are flushed to their backing files.
+ 
+## Snapshot versioning
+
+Firecracker snapshotting implementation offers support for microVM versioning
+(`cross-version snapshots`) in the following contexts:
+- saving snapshots at older versions (also called `rollback`),
+- loading snapshots from older versions (also called `live update`).
+Live update refers to successfully loading any snapshot created by Firecracker version
+`N` in a Firecracker version `N+o`.
+Rollback means being able to create a snapshot with any version in `[N, N + o]` interval
+while being in Firecracker version `N+o`.
+
+The value of `o` (maximum number of older versions that we can restore from / save
+a snapshot to, from the current version) is not yet decided, but we plan on offering
+backwards compatibility for at least one year changes. 
```

Comment:
This sounds like a promise and we don't have anything decided yet. Please remove.